### PR TITLE
gpu: Replace shaders if layout unmodified

### PR DESCRIPTION
**Pull request by Tony-LunarG**
If all descriptor sets are used by the app, gpu-av doesn't modify the pipeline layout, and in this case it should remove the instrumented shaders at pipeline creation

### CODE REVIEW

File: layers/gpu_validation.cpp [hunk 1]
```diff
@@ -427,28 +427,32 @@ std::vector<safe_VkGraphicsPipelineCreateInfo> CoreChecks::GpuPreCallRecordCreat
...
-    if (std::all_of(pipeline_uses_debug_index.begin(), pipeline_uses_debug_index.end(), [](unsigned int i) { return i == 0; })) {
+    // See if any pipeline has shaders using the debug descriptor set index or if its layout hasn't been modified
+    if (std::all_of(replace_shaders.begin(), replace_shaders.end(), [](bool i) { return i == false; })) {
         // None of the shaders in all the pipelines use the debug descriptor set index, so use the pipelines
         // as they stand with the instrumented shaders.
         return new_pipeline_create_infos;
     }
 
-    // At least one pipeline has a shader that uses the debug descriptor set index.
+    // At least one pipeline has a shader that uses the debug descriptor set index or its layout hasn't been modified
```
Comment (jzulauf-lunarg) [inline layers/gpu_validation.cpp]: Some clarifying comments would be useful.

File: layers/gpu_validation.cpp [hunk 2]
```diff
@@ -427,28 +427,32 @@ std::vector<safe_VkGraphicsPipelineCreateInfo> CoreChecks::GpuPreCallRecordCreat
...
     for (uint32_t pipeline = 0; pipeline < count; ++pipeline) {
-        if (pipeline_uses_debug_index[pipeline]) {
+        if (replace_shaders[pipeline]) {
```
Comment (jzulauf-lunarg) [inline layers/gpu_validation.cpp]: feels like this could be folded into the loop above .

File: layers/gpu_validation.cpp [hunk 3]
```diff
@@ -427,28 +427,32 @@ std::vector<safe_VkGraphicsPipelineCreateInfo> CoreChecks::GpuPreCallRecordCreat
...
+        // If the app requests all available sets, the pipeline layout was never modified and the shaders need to be replaced
+        if (pipe_state[pipeline]->pipeline_layout.set_layouts.size() >= gpu_validation_state->adjusted_max_desc_sets) {
+            replace_shaders[pipeline] = true;
         }
     }
 
-    // See if any pipeline has shaders using the debug descriptor set index
-    if (std::all_of(pipeline_uses_debug_index.begin(), pipeline_uses_debug_index.end(), [](unsigned int i) { return i == 0; })) {
+    // See if any pipeline has shaders using the debug descriptor set index or if its layout hasn't been modified
+    if (std::all_of(replace_shaders.begin(), replace_shaders.end(), [](bool i) { return i == false; })) {
```
Comment (jzulauf-lunarg) [inline layers/gpu_validation.cpp]: I don't believe this is saving any performance or iterations...

File: layers/gpu_validation.cpp (hunk 1 above)
Comment (Tony-LunarG) [reply layers/gpu_validation.cpp]: "If the app requests all available sets, the pipeline layout was not modified at pipeline layout creation and the already instrumented shaders need to be replaced with uninstrumented shaders"
added to number of sets check

File: layers/gpu_validation.cpp (hunk 2 above)
Comment (Tony-LunarG) [reply layers/gpu_validation.cpp]: Folded

File: layers/gpu_validation.cpp (hunk 3 above)
Comment (Tony-LunarG) [reply layers/gpu_validation.cpp]: Folded into one loop

**Pull request merged**